feat(sourcemapprocessor): enhance symbolication process with per-stacktrace Error Caching

clinton-slice opened the pull request:
<!--
Thank you for contributing to the project! 💜
Please see our [OSS process document](https://github.com/honeycombio/home/blob/main/honeycomb-oss-lifecycle-and-practices.md#) to get an idea of how we operate.
-->

## Which problem is this PR solving?

- Closes: [FO-599](https://linear.app/honeycombio/issue/FO-599/per-stacktrace-deduping-sourcemapprocessor)

## Short description of the changes

Added per-stacktrace error caching to `sourcemapprocessor` to prevent repeatedly fetching missing source maps (404s, timeouts, etc.) within a single stacktrace.

- Introduced error caching for failed symbolication attempts to avoid redundant fetches for URLs without source maps.
- Updated the symbolication logic to return mapped values based on line and column adjustments, simulating real source map behavior.
- Added unit tests to verify deduplication of symbolication calls and ensure unique results for frames with the same URL but different line/column positions.

**Logic**:
1. Before symbolicating a frame, check if we've seen this URL fail before in this stacktrace
2. If yes, reuse the cached error (avoiding redundant failed fetches)
3. If no, perform symbolication and cache only errors
4. Successful symbolications are performed for each frame (since line/column position matters)
5. Source maps themselves are cached separately in `basicSymbolicator.cache` (LRU cache)
6. Cache is automatically GC'd after stacktrace processing completes

## How to verify that this has the expected result

---

- [x] CHANGELOG is updated
- [ ] README is updated with documentation

clinton-slice committed:
feat: enhance symbolication process with error caching and improved mapping

- Introduced error caching for failed symbolication attempts to avoid redundant fetches for URLs without source maps.
- Updated the symbolication logic to return mapped values based on line and column adjustments, simulating real source map behavior.
- Added unit tests to verify deduplication of symbolication calls and ensure unique results for frames with the same URL but different line/column positions.

clinton-slice committed:
chore: update CHANGELOG to reflect performance improvements in source map fetching

- Added entry for performance enhancement that reduces redundant failed source map fetch attempts by 80-95% through error caching. | @clintonnkemdilim

jairo-mendoza commented on sourcemapprocessor/CHANGELOG.md:
```diff
@@ -2,6 +2,8 @@
 
 ## Unreleased
 
+- perf: reduce redundant failed source map fetch attempts by 80-95% per stacktrace through error caching | @clintonnkemdilim
```
nit: could we change this line to what we have in this pr's title?

clinton-slice replied:
Thanks. I have changed this

clinton-slice committed:
feat: enhance symbolication process to handle validation errors

- Added a new test to ensure that validation errors for individual frames do not prevent subsequent valid frames from being symbolicated.
- Updated the symbolicator logic to clarify that frame-specific validation errors are not cached, allowing valid frames to still attempt symbolication even after an invalid frame.

clinton-slice committed:
chore: update CHANGELOG to include performance enhancement for symbolication process with error caching | @clintonnkemdilim

clinton-slice committed:
chore: update CHANGELOG to reflect performance enhancements in symbolication process with error caching | @clintonnkemdilim

clinton-slice committed:
feat: implement cache key generation for symbolication process

- Added a new function `buildCacheKey` to create consistent cache keys for source maps, incorporating the build UUID when present.
- Updated the symbolicator logic to utilize the new cache key format, ensuring separate cache entries for different builds.
- Introduced unit tests to validate the cache key generation and its impact on cache behavior.

beekhc commented on sourcemapprocessor/trace_processor.go:
```diff
@@ -173,11 +173,44 @@ func (sp *symbolicatorProcessor) processThrow(ctx context.Context, attributes pc
+
+			// Cache only errors - successful results vary by line/column so can't be reused
+			if err != nil {
+				errorCache[cacheKey] = err
```
It's kind of a code smell to repeat the validation here. But more importantly, I'm not convinced the new code is correct. For example, we propagate errors from `symbolic.NewSourceMapCache`, and those aren't necessarily fetch errors, but would be treated as such here.

I think we're gonna need to change `sp.symbolicator.symbolicate` to indicate whether the failure was a fetch error. I think the most idiomatic way to do this in Go is, for fetch errors, to return an error with a specific type that wraps the underlying error. Any non-fetch error can be returned as-is. Then, in the calling code, you would check use `errors.Is` to check if the failure was a fetch error.

https://pkg.go.dev/errors#Is

beekhc replied:
To be more specific, I think it's this particular return where we should be wrapping the err with a new type:

https://github.com/honeycombio/opentelemetry-collector-symbolicator/blob/047e87af1d390244cbc0d6bf1ebbbe4dbca06b6c/sourcemapprocessor/symbolicator.go#L109

clinton-slice replied:
Its been implemented. Thanks @beekhc

clinton-slice committed:
feat: introduce FetchError type for improved error handling in symbolication

- Added a new FetchError type to wrap source map fetch failures, providing better context for errors like 404 and timeout.
- Updated the symbolicator to return FetchError instances instead of generic errors for fetch failures.
- Modified tests to handle FetchError, ensuring that errors are correctly cached and reused for fetch failures while maintaining the behavior for validation errors.
- Enhanced comments to clarify caching behavior for different error types in the symbolication process.

beekhc reviewed (approved):
This looks good to me!

jairo-mendoza reviewed (approved):
this looks solid to me, i just have one nit. thank you for adding the tests!

clinton-slice committed:
refactor: rename errorCache to fetchErrorCache for clarity in symbolication process

- Updated variable name from `errorCache` to `fetchErrorCache` to better reflect its purpose of caching fetch errors specifically.
- Adjusted related logic to ensure consistency with the new naming convention, enhancing code readability and maintainability.

clinton-slice committed:
refactor: simplify comments on fetch error caching in symbolication process
